pygfx/utils/transform.py

```diff
@@ -0,0 +1,244 @@
+import numpy as np
+import pylinalg as pla
+from datetime import datetime
+
+from typing import List
+
+
+class cached:
+    """Cache for computed properties.
+
+    This descriptor implements a minimal timestamp-based cache for computed
+    properties. The value of the property is computed using ``update_fn`` and
+    the result is cached until ``obj.last_modified`` advances. At this point the
+    value of the computed property is recomputed upon the next read.
+
+
+
+    Example
+    -------
+
+    class Foobar:
+        def __init__(self, x):
+            self._value = 0
+            self.last_modified = datetime.now()
+
+        @property
+        def value(self):
+            return self._value
+
+        @value.setter
+        def value(self, new_value):
+            self._value = new_value
+            self.last_modified = datetime.now()
+
+        @property
+        @cached
+        def double(self):
+            print("Computed value of double..")
+            return 2*x
+
+    """
+
+    def __init__(self, update_fn=None) -> None:
+        self.update_fn = update_fn
+        self.last_updated = None
+        self.cached_value = None
+
+    def __get__(self, obj: "AffineBase", objtype=None):
+        if obj is None:
+            return self
+
+        if self.last_updated is None or obj.last_modified > self.last_updated:
+            self.cached_value = self.update_fn(obj)
+            self.last_updated = obj.last_modified
+
+        return self.cached_value
+
+    def __call__(self, obj):
+        """Helper to allow interoperability with @property."""
+        return self.__get__(obj)
+
+
+class AffineBase:
+    last_modified: datetime
```

Comment:
is this just a type annotation or a class attribute? descriptors clearly come naturally to you :) I'm still developing my intuitive understanding of them

Reply:
This is just a type annotation. Each class must define `last_modified` for the cache to work, but for some subclasses (`LinkedTransform`) this is better implemented using a computed `@property`, and for others (`AffineTransform`) via an actual attribute. It still has to exist though, so I declare it on the base to get type hints and auto-completion.

I'm actually not 100% happy with the current version of the implementation yet, because `last_modified` is a magic value that the descriptor assumes exists. Once I figure out a way to make this less magical I will refactor.


I can also see where this confusion comes from since it would be an attribute for a `@dataclass`, i.e. if I did

```python
@dataclass
class AffineBase:
    last_modified: datetime
```

then it would be a (required) attribute. This works, because `@dataclass` adds custom logic to `__new__` which creates the fields from the provided type annotation.

> descriptors clearly come naturally to you :)

They are a bit of a weird idea. I wrote an [SDF parser ](https://github.com/FirefoxMetzger/python-sdformat) during my PhD that uses them excessively (SDF is a special flavor of XML), and that's where my understanding of them comes from.

The intuitive way to think of them is as a mechanism to implement getter/setter logic that can be reused in different places (across classes and attributes). `__get__` holds the getter logic, `__set__` holds the setter logic. The "trick" is to not get confused by what `self` refers to inside the descriptor class so that there is no confusion as to where the data is stored.

If we want to create a read-only element, then we could use `@property` and not implement a setter. We would store the private value on the instance of the class (`class_instance`) and typically do something like this:

```python
class Simple:
    def __init__(self):
        self._gettable = 42

    @property
    def gettable(self):
        return self._gettable
```

We can do the same with descriptors but choose to store the value on the instance of the descriptor (`descriptor_instance`) instead of the class:

```python
class ReadOnly:
    # Note: I'm renaming `self` to `descriptor_instance` here to avoid confusion
    def __init__(descriptor_instance, value):
        descriptor_instance._gettable = value

    def __get__(descriptor_instance, class_instance, objtype=None):
        # look for the value on the descriptor
        value = descriptor_instance._gettable
        return value

class Simple:
    gettable = ReadOnly(42)
```

Both implementations have the same effect (`gettable` is read-only), but `Simple` no longer has `Simple._gettable` which makes it much harder to accidentally modify the read-only value.

Alternatively, we could use the descriptor to place `_gettable` on the `class_instance`. However, this can cause a name collision if I use the descriptor multiple times on the same class (each instance of the descriptor has its own `descriptor_instance._gettable`, but they all share `class_instance._gettable`). To de-alias this, we have `__set_name__`, which is called with the variable name to which `descriptor_instance` is assigned. With it, we can re-implement the same behavior as with `@property`:

```python
class ReadOnly:
    def __init__(descriptor_instance, value):
        descriptor_instance.value = value
        descriptor_instance.name = None

    def __set_name__(descriptor_instance, clazz, name):
        storage_name = "_"+name
        descriptor_instance.name = storage_name  # remember where we stored the value
        setattr(clazz, storage_name, descriptor_instance.value)

    def __get__(descriptor_instance, class_instance, objtype=None):
        # look for the value on the class instance
        value = getattr(class_instance, self.name)
        return value

class Simple:
    gettable = ReadOnly(42)
    foo = ReadOnly("bar")
```

Now `Simple` has both `Simple.gettable` and `Simple._gettable` (similar to the `@property` approach), as well as `Simple.foo` and `Simple._foo`.

The `@cached` decorator syntax that I use in the PR is just a flourish. Decorators call the decorating function using the decorated function as input. In the case of `@property` or `@cached` the decorating function is the constructor (i.e., `cached.__init__(self, update_fn)`) which is called using `cached(<handle_of_decorated_function>)` or `cached(update_fn=<handle_of_decorated_function>)`.

In the same way, we need `cache.__call__` to make the descriptor "stackable" with `@property`, which expects something callable as its first input to `property.__init__(self, callable)`.

Does that help?